src/dataflow-types/src/client/controller/storage.rs

```diff
@@ -444,14 +447,15 @@ where
                     .as_option()
                     .map(|ts| ts.clone().try_into().expect("negative timestamp")),
             );
-            let upper = ts_binding_collection.upper()?;
+            let upper = self.state.stash.upper(ts_binding_collection)?;
             // TODO(petrosagg): This guard should go away by ensuring storage workers never re-send
             // the bindings and frontiers they were initialized with
             if PartialOrder::less_than(&upper, &seal_frontier) {
-                ts_binding_collection.seal(seal_frontier.borrow())?;
+                seals.push((ts_binding_collection, seal_frontier));
             }
             durability_updates.push((*id, write_frontier));
         }
+        self.state.stash.seal_batch(&seals)?;
```

Comment:
In light of Dan's comment that we might not be able to provide atomic operations depending on the stash impl, is it safe for this to partially apply?

Reply:
Yes, these methods are explicitly provided as merely a performance optimization and make no atomicity guarantees

Reply:
I mean, is it safe for the timestamp bindings if there's an error mid batch?

Reply:
Yes, currently we produce timestamp bindings from a single writer and any error will eventually propagate up to `materialized` and crash the process which will force a clean resumption from whatever state we left of. We can become a bit more sophisticated with retries and such when we get an idempotent `append` API call